lms/util/h.py

```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+
+"""Utility functions for working with the h API."""
+
+from __future__ import unicode_literals
+
+import random
+import string
+
+
+USERNAME_ALLOWED_CHARS = string.ascii_letters + string.digits + "._"
+"""The characters that are allowed in h usernames."""
+
+
+DISPLAY_NAME_MAX_LENGTH = 30
+"""The maximum length of an h display name."""
+
+
+USERNAME_MAX_LENGTH = 30
+"""The maximum length of an h username."""
+
+
+def generate_display_name(request_params):
+    full_name = (request_params.get("lis_person_name_full") or "").strip()
+    given_name = (request_params.get("lis_person_name_given") or "").strip()
+    family_name = (request_params.get("lis_person_name_family") or "").strip()
+
+    if full_name:
+        display_name = full_name
+    else:
+        display_name = " ".join((given_name, family_name))
+
+    display_name = display_name.strip()
+
+    display_name = display_name or "Anonymous"
+
+    if len(display_name) > DISPLAY_NAME_MAX_LENGTH:
+        display_name = display_name[:DISPLAY_NAME_MAX_LENGTH - 1].rstrip() + "…"
```

Comment:
Curious: what's the `rstrip` for (if your'e doing it _after_ you slice the string)?

Reply:
I didn't put in a specific test for this, but it is incidentally covered by one of the other tests: If the display name is `"Firstname Lastname"` and the truncation happens to cut if off after the space (`"Firstname "`) then you end up with a trailing space (you would get `"Firstname …"` instead of `"Firstname…"`).